internal/dyngo/events.go

```diff
@@ -0,0 +1,220 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package dyngo
+
+import (
+	"reflect"
+	"sync"
+	"sync/atomic"
+)
+
+type (
+	eventManager struct {
+		onStart, onData, onFinish eventRegister
+
+		disabled bool
+		mu       sync.RWMutex
+	}
+
+	eventRegister struct {
+		mu        sync.RWMutex
+		listeners eventListenerMap
+	}
+
+	// eventListenerMap is the map of event listeners. The list of listeners are indexed by the operation argument or
+	// result type the event listener expects. The map key type is a struct of the fully-qualified type name (type
+	// package path and name) to uniquely identify the argument/result type instead of using the interface type
+	// reflect.Type that can possibly lead to panics if the actual value isn't hashable as required by map keys.
+	eventListenerMap map[eventListenerMapKey][]eventListenerMapEntry
+
+	eventListenerMapKey struct {
+		pkgPath, name string
+	}
+
+	eventListenerMapEntry struct {
+		Id       eventListenerID
+		Callback interface{}
+	}
+)
+
+type (
+	EventListenerID interface {
+		unregisterFrom(*Operation)
+	}
```

Comment:
I don't think polymorphism is required for this. `EventListenerID` is not really an ID, but a way to provide different unregistration logic depending on which event the listener was registered to.

We can just return a function `func (*Operation)` which performs the unregistration rather than having separate types. This has the added benefit of keeping the registration/unregistration logic in the same place.

Reply:
Great idea! It's now implemented in 335156f